feat: implement Caching API

jamagalhaes opened the pull request:
[KOALA-2746](https://canvasmedical.atlassian.net/browse/KOALA-2746)

[KOALA-2746]: https://canvasmedical.atlassian.net/browse/KOALA-2746?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

jamagalhaes commented:
@aduane / @beaugunderson 

With this implementation, I aimed to follow a hybrid API design—one that can be used both internally within the SDK (for future use cases) and externally by plugins.

To expose this same API to plugins safely, I had to update the sandbox rules to prevent access to private properties. This ensures that plugin developers can’t directly access the underlying cache connection or perform unintended operations.

I realize this introduces a breaking change: from now on, access to attributes starting with _ will be restricted. This aligns with the default behavior of RestrictedPython's getattr, which we had previously overridden.

If you're not comfortable with this approach, I'm open to discuss different naming patterns to restrict, i.e. double underscore, or `__internal__`, etc

beaugunderson commented:
cc @djantzen for security improvement

djantzen commented:
I'm in favor of the change, just wondering if we have plugins in the wild that are using forbidden methods. I think I overrode `getattr` in order not to break existing workflow kit protocols when I introduced RP there.

djantzen commented on canvas_sdk/caching/utils.py:
```diff
@@ -0,0 +1,25 @@
+        """
+        if hasattr(obj, self.private_name):
+            raise AttributeError(f"{self.public_name} cannot be changed.")
+        setattr(obj, self.private_name, value)
```
Probably, but why do we want to prevent them from overwriting their own values if they want to?

beaugunderson replied:
this is for security, not the cache values themselves

jamagalhaes replied:
@beaugunderson which class exactly? they are only allowed to import get_client from `.plugins` module. They won't be able to subclass `Cache`, unless I'm missing something 🤔

djantzen replied:
Are we intending for this to be PHI-safe storage mechanism? I see that we are restricting cache access to the plugin itself, but if customers are allowed or encouraged to store PHI here it seems like the risk goes way up for data leakage or corruption.

djantzen replied:
Ok, which Redis will this be pointed to? I'm looking at `training` for comparison, and we do not use SSL for connections to Redis from the home-app. Difference in connection string is `redis://` versus `rediss://`. Of course, none of our other usage of Redis AFAIK store PHI or PII.

beaugunderson replied:
we should be using SSL for all redis connections at all times, whether it's for caching or for channel layers it will contain PHI (how could it not?)

djantzen replied:
Should be, yes; are we? The URLs we specify in the environment are non-SSL. What do you mean by channel layers? Is that for websockets?

beaugunderson replied:
it provides websocket support but we serve Django via [django-channels](https://channels.readthedocs.io/en/latest/)... in `app` we go through a `ProtocolTypeRouter` from channels that routes traffic based on protocol, events are sent over a redis bus

we need to remediate the non-SSL URLs as soon as safely possible (meaning we should test them first on a non-prod environment)

jamagalhaes commented:
> I'm in favor of the change, just wondering if we have plugins in the wild that are using forbidden methods. I think I overrode `getattr` in order not to break existing workflow kit protocols when I introduced RP there.

In addition to @beaugunderson  suggestions, I enhanced the `__getattr__` safeguard to allow access to private attributes defined within the same package as the plugin. I also refactored the `__write__` guard to explicitly prevent assignments to PROTECTED_RESOURCES

djantzen commented on canvas_sdk/caching/client.py:
```diff
@@ -0,0 +1,21 @@
+from django.core.cache import InvalidCacheBackendError
+from django.core.cache import caches as django_caches
```
Might be worth writing a test to ensure that this cannot be loaded into a plugin. It should fail currently as this is not in the allow list of modules, but it would be some insurance against a change making it available at some future time. I'm suggesting it because it seems like if an attacker gains access to the cache itself then they can bypass the plugin-specific segmentation of the data.

jamagalhaes replied:
@djantzen done

djantzen replied:
Thanks!

beaugunderson commented on plugin_runner/sandbox.py:
```diff
@@ -206,13 +252,10 @@ def check_for_name_in_iterable(self, iterable_node: ast.Tuple | ast.List) -> Non
 
         def visit_Attribute(self, node: ast.Attribute) -> ast.AST:
             """Checks and mutates attribute access/assignment.
```
why delete the vertical whitespace? previous was more readable

beaugunderson reviewed (approved):
looking good minus the nitpick about vertical whitespace!

@djantzen can you follow up when the plugin-specific redis is created?

aduane commented:
Ticket to track the work needed for plugin-specific redis instance: https://canvasmedical.atlassian.net/browse/KOALA-2779

djantzen commented:
The more I think about this the more uneasy I am. This is the first time we've allowed customer data to co-mingle in a data store in a mutable state. And plugins have the ability to mutate it and read it, which means that we're putting all our trust that PHI won't be exposed or corrupted into a single annotation that namespaces the data. 

My POV here may be colored by the fact that I'm presently a claimant in two separate class action lawsuits about mishandling medical PHI. But I don't think we should do this unless we have a per-instance or per-customer encryption key such that in no case can one customer intercept another's. Even that would leave open the possibility of writing bad data into another customer's namespace.

Are we doing this with Redis for performance reasons, and we truly need an in-memory data store? Or are we doing it because we haven't gamed out using the instance's dedicated Postgres for direct control by customer code? Providing this capability using Postgres would ensure against the risks of leakage and corruption. It opens up new questions for sure, but I believe we're going to want to answer those questions anyway long term, as customers will want to model more and more sophisticated processes and store state permanently, as they do in other platforms such as Salesforce.

beaugunderson commented on canvas_sdk/caching/plugins.py:
```diff
@@ -0,0 +1,13 @@
+def get_cache(**kwargs: Any) -> Cache:
+    """Get the cache client for plugins."""
+    prefix = kwargs["plugin_name"]
+    return get_cache_client("plugins", prefix, settings.CANVAS_SDK_CACHE_TIMEOUT_SECONDS)
```
`prefix` here is the plugin name which is not globally unique... this needs to use `CUSTOMER_IDENTIFIER` as well!

per @djantzen's point this would lead to potential corruption as soon as two customers upload `my-first-plugin` and modify it to use caching.

let's talk through the implications here in planning on Tuesday (though that may get bumped if we run out of time since we already have a topic for Tuesday). we could also meet separately to discuss.

jamagalhaes replied:
@beaugunderson check `CACHES` configuration at `settings.py`, this is already enforced there.

beaugunderson replied:
ah, I saw that in my original review but not the re-review 😆 

so the effective collision space is `CUSTOMER_IDENTIFIER/PLUGIN_NAME` which is very constrained.

the redis cache methods all call `make_and_validate_key` which call `make_key` and `validate_key` in turn. `validate_key` is a misnomer, all it does is validation for memcached which we're not using.

`make_key` uses `self.key_func`, which by default respects the `key_prefix` that is passed in, and we don't override it.

so the question is, can someone change the key prefix at runtime? I will take a stab at it.

beaugunderson commented:
just added the latest sandbox improvements; going to rebase now as well.

jamagalhaes commented:
@beaugunderson why `mocker.stopall()` and `nuke_cache()` were necessary?
